webrender_traits/src/types.rs

```diff
@@ -322,18 +330,62 @@ pub struct GlyphKey {
     pub size: Au,
     pub index: u32,
     pub color: ColorU,
+    pub x_suboffset: SubpixelOffset,
+    pub y_suboffset: SubpixelOffset,
+    pub render_mode: FontRenderMode,
+    pub glyph_options: Option<GlyphOptions>,
 }
 
 impl GlyphKey {
     pub fn new(font_key: FontKey,
                size: Au,
                color: ColorF,
-               index: u32) -> GlyphKey {
-        GlyphKey {
+               index: u32,
+               x: f32,
+               y: f32,
+               render_mode: FontRenderMode,
+               glyph_options: Option<GlyphOptions>) -> GlyphKey {
+        let mut key = GlyphKey {
             font_key: font_key,
             size: size,
             color: ColorU::from(color),
             index: index,
+            x_suboffset: SubpixelOffset::Zero,
+            y_suboffset: SubpixelOffset::Zero,
+            render_mode: render_mode,
+            glyph_options: glyph_options,
+        };
+
+        key.set_x_offset(x);
+        key.set_y_offset(y);
+
+        key
+    }
+
+    pub fn set_x_offset(&mut self, x: f32) {
+        self.x_suboffset = self.subpixel_quantize_offset(x);
+    }
+
+    pub fn set_y_offset(&mut self, y: f32) {
+        self.y_suboffset = self.subpixel_quantize_offset(y);
+    }
+
+    // Skia quantizes subpixel offets into 1/4 increments.
+    // Given the absolute position, return the quantized increment
+    fn subpixel_quantize_offset(&self, pos: f32) -> SubpixelOffset {
```

Comment:
Let's make this a free function that takes a render mode as a parameter. Then the constructor for GlyphKey can be simplified to call this directly while creating the struct.